### PR TITLE
Implement UUPS proxy (ERC1822)

### LINKED ISSUE #2421: Provide implementation of UUPS proxy pattern
[EIP 1822](https://eips.ethereum.org/EIPS/eip-1822) describes the UUPS proxy pattern. This pattern is simpler and should be cheaper as well, so we're interested in providing it as an alternative to the transparent proxy pattern.

### PR DESCRIPTION
**Pull request by Amxx**
Fixes #2421

Renames `UpgradeableProxy` into `UUPSProxy`, with the proxy only containing the initializing constructor and the `_implementation()` function. UUPS slot related tools are in a `UUPS` abstract contract, used by both the `UUPSProxy` and the `UUPSProxiable` contracts.

Contract that used to inherit from `UpgradeableProxy` now inherit from `UUPSProxy`, with no further changes needed.

#### PR Checklist
- [x] Tests
- [ ] Documentation
- [ ] Changelog entry

### CODE REVIEW

**Comment by Amxx** [comment]
Thoughts:

- `_upgradeTo(addrees) internal` and `_setImplementation(addrees) private` might not both be needed. Should there be only one (internal) function that sets + emit event ? 
- should there be a `_readImplementation() internal` that is used in `UUPSProxy._implementation()` ?

File: contracts/proxy/beacon/BeaconProxy.sol [hunk 1]
```diff
@@ -34,55 +28,13 @@ contract BeaconProxy is Proxy {
      */
     constructor(address beacon, bytes memory data) payable {
         assert(_BEACON_SLOT == bytes32(uint256(keccak256("eip1967.proxy.beacon")) - 1));
-        _setBeacon(beacon, data);
-    }
-
-    /**
-     * @dev Returns the current beacon address.
-     */
-    function _beacon() internal view virtual returns (address beacon) {
```
Comment (frangio) [inline contracts/proxy/beacon/BeaconProxy.sol:43]: We need to keep this one as an alias to `_getBeacon`.

File: contracts/proxy/README.adoc [hunk 2]
```diff
@@ -9,12 +9,29 @@ The abstract {Proxy} contract implements the core delegation functionality. If t
...
 
+= UUPS design: upgradeability as an implementation feature
+
+Upgradeable smart contracts rely on proxies to relay the calls in a way that is programmable. As discussed previously, we provide different proxy contracts that each come with a specific set of features. Other designs, not (yet) proposed as part of the OpenZeppelin products, also exist.
+
+The most simple, and common, design is known as Universal Upgradeable Proxy Standard (UUPS). This design is both lightweight and versatile and is proposed through the {ERC1967Proxy} and the {Proxiable} contract.
+
+While {Proxiable} uses the same interface as {TransparentUpgradeableProxy}, in the first case the upgrade is handled by the implementation, and can eventually be removed. {TransparentUpgradeableProxy}, on the other hand, includes the upgrade logic in the proxy. This means {TransparentUpgradeableProxy} is more expensive to deploy. Note that, since both proxies use the same storage slot for the implementation address, using a UUPS compliant implementation with a {TransparentUpgradeableProxy} might allow non-admins to perform upgrade operations.
```
Comment (frangio) [inline contracts/proxy/README.adoc:24]: Replace mentions of `Proxiable` with `UUPSUpgradeable` in this file.

File: contracts/proxy/ERC1967/ERC1967Upgrade.sol [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+     * Emits an {Upgraded} event.
+     */
+    function _upgradeToAndCallSecure(address newImplementation, bytes memory data) internal {
+        _upgradeToAndCallSecure(newImplementation, data, false);
+    }
+
+    function _upgradeToAndCallSecure(address newImplementation, bytes memory data, bool forceCall) internal {
+        address oldImplementation = _getImplementation();
+        // check if nested in an upgrade check
+        StorageSlot.BooleanSlot storage upgradePending = StorageSlot.getBooleanSlot(_UPGRADE_PENDING_SLOT);
```
Comment (frangio) [inline contracts/proxy/ERC1967/ERC1967Upgrade.sol:61]: Can we move this declaration right before `if (!upgradePending.value)`?

File: contracts/proxy/ERC1967/ERC1967Proxy.sol [hunk 4]
```diff
@@ -23,56 +20,13 @@ contract ERC1967Proxy is Proxy {
...
-            impl := sload(slot)
-        }
-    }
-
-    /**
-     * @dev Upgrades the proxy to a new implementation.
-     *
-     * Emits an {Upgraded} event.
-     */
-    function _upgradeTo(address newImplementation) internal virtual {
```
[the remark this answers is not in the document]
Comment (Amxx) [reply contracts/proxy/ERC1967/ERC1967Proxy.sol:60]: added _upgradeTo to ERC1967Upgrade

**Review by frangio — COMMENTED** [review]
Two things before merging:

1. Removing the `_upgradeToAndCall*` overloads, keeping only the version with the `forceCall` argument.
2. Adding a test that an upgrade to a non-UUPSUpgradeable contract reverts.

**Review by frangio — APPROVED** [review]
LGTM. Thank you.

**Pull request merged**